astropy/time/core.py

```diff
@@ -276,11 +266,57 @@ def new_like(self, cols, length, metadata_conflicts='warn', name=None):
         return out
 
 
-class TimeDeltaInfo(TimeInfo):
-    _represent_as_dict_extra_attrs = ('format', 'scale')
+class TimeInfo(TimeInfoBase):
+    def _represent_as_dict(self, attrs=None):
+        """Get the values for the parent ``attrs`` and return as a dict.
+
+        By default, uses '_represent_as_dict_attrs'.
+        """
+        map = super()._represent_as_dict(attrs=attrs)
+
+        # The datetime64 format requires special handling for ECSV (see #12840).
+        # The `value` has numpy dtype datetime64 but this is not an allowed
+        # datatype for ECSV. Instead convert to a string representation.
+        if (self._serialize_context == 'ecsv'
+                and map['format'] == 'datetime64'
+                and 'value' in map):
+            map['value'] = map['value'].astype('U')
+
+        # The datetime format is serialized as ISO with no loss of precision.
+        if map['format'] == 'datetime':
+            map['value'] = np.array([val.isoformat() for val in map['value']])
```

Comment:
This won't work for multi-dimensional arrays. Note that my suggestion of `.astype('U')` is equivalent to an array-shape keeping application of `str` - it seems to give exactly the same output for 1-D except that it is with a space instead of a T. If you do want 'isot', then `np.vectorize(lambda x: x.isoformat())(t.value)` would work.
...
Hmm, somewhat surprisingly, `vectorize` is faster than `.astype('U')`, and in fact very similar to your array call -- I suggest to go with it!
```
t = Time(np.linspace(50000., 50001., 8000), format='mjd')
value = t.datetime
%timeit np.vectorize(lambda x: x.isoformat())(value)
10.8 ms ± 411 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit value.astype('U')
21.6 ms ± 353 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit np.array([_.isoformat() for _ in value])
10.2 ms ± 170 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```